state/storage.go

```diff
@@ -819,32 +800,45 @@ func (s *Storage) RoomMembershipDelta(roomID string, from, to int64, limit int)
 }
 
 // Extract all rooms with joined members, and include the joined user list. Requires a prepared snapshot in order to be called.
-func (s *Storage) AllJoinedMembers(txn *sqlx.Tx, tempTableName string) (result map[string][]string, metadata map[string]internal.RoomMetadata, err error) {
+func (s *Storage) AllJoinedMembers(txn *sqlx.Tx, tempTableName string) (joinedMembers map[string][]string, metadata map[string]internal.RoomMetadata, err error) {
 	rows, err := txn.Query(
-		`SELECT room_id, state_key from ` + tempTableName + ` INNER JOIN syncv3_events on membership_nid = event_nid WHERE membership='join' OR membership='_join' ORDER BY event_nid ASC`,
+		`SELECT room_id, state_key, membership from ` + tempTableName + ` INNER JOIN syncv3_events
+		on membership_nid = event_nid WHERE membership='join' OR membership='_join' OR membership='invite' OR membership='_invite' ORDER BY event_nid ASC`,
```

Comment:
In the deleted query you additionally had the condition `event_type='m.room.member'` in the join clause. Does adding this here make any difference to the query plan or execution time?

Reply:
Negatory, it just changes the filter.
```
syncv3=# EXPLAIN SELECT room_id, state_key, membership from x INNER JOIN syncv3_events on membership_nid = event_nid WHERE membership='join' OR membership='_join' OR membership='invite' OR membership='_invite' ORDER BY event_nid ASC;
                                                                          QUERY PLAN                                                                          
--------------------------------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=18913.61..19049.28 rows=54266 width=67)
   Sort Key: syncv3_events.event_nid
   ->  Hash Join  (cost=10451.18..12420.20 rows=54266 width=67)
         Hash Cond: (x.membership_nid = syncv3_events.event_nid)
         ->  Seq Scan on x  (cost=0.00..1665.86 rows=115486 width=8)
         ->  Hash  (cost=9768.74..9768.74 rows=54595 width=67)
               ->  Seq Scan on syncv3_events  (cost=0.00..9768.74 rows=54595 width=67)
                     Filter: ((membership = 'join'::text) OR (membership = '_join'::text) OR (membership = 'invite'::text) OR (membership = '_invite'::text))
(8 rows)

syncv3=# EXPLAIN SELECT room_id, state_key, membership from x INNER JOIN syncv3_events on membership_nid = event_nid WHERE event_type='m.room.member' AND  (membership='join' OR membership='_join' OR membership='invite' OR membership='_invite') ORDER BY event_nid ASC;
                                                                                               QUERY PLAN                                                                                                
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=19177.89..19313.07 rows=54071 width=67)
   Sort Key: syncv3_events.event_nid
   ->  Hash Join  (cost=10739.19..12708.22 rows=54071 width=67)
         Hash Cond: (x.membership_nid = syncv3_events.event_nid)
         ->  Seq Scan on x  (cost=0.00..1665.86 rows=115486 width=8)
         ->  Hash  (cost=10059.21..10059.21 rows=54399 width=67)
               ->  Seq Scan on syncv3_events  (cost=0.00..10059.21 rows=54399 width=67)
                     Filter: ((event_type = 'm.room.member'::text) AND ((membership = 'join'::text) OR (membership = '_join'::text) OR (membership = 'invite'::text) OR (membership = '_invite'::text)))
```